File: src/lib/service-contract/workspace.ts
import type {
  EquityPoint,
  ImportPreflightSeverity,
  ImportPreflightStatus,
  OperationStatus,
  OrchestratorMode
} from "./common";

export type WorkspaceCatalogCategory =
  | "entrypoint"
  | "active"
  | "index"
  | "export"
  | "operation"
  | "checkpoint"
  | "agent"
  | "environment"
  | "adapter"
  | "collection"
  | "entry"
  | "import"
  | "blob"
  | "session"
  | "evaluation";

export type CheckpointSummary = {
  id: string;
  alias: string;
  type: "promotion" | "export" | "incident";
  typeTone: "positive" | "warning" | "danger";
  summary: string;
  createdAt: string;
  performance: string;
  pathRef: string;
  exportBundleRef?: string;
};

export type ExportBundleState = {
  exportId: string;
  createdAt: string;
  policyId: string;
  checkpointRef: string;
  workspaceRef: string;
  bundleRef: string;
  includedRefs: string[];
  excludedPaths: string[];
  sanitized: boolean;
};

export type CheckpointDetailState = {
  id: string;
  alias: string;
  type: "promotion" | "export" | "incident";
  typeTone: "positive" | "warning" | "danger";
  summary: string;
  createdAt: string;
  performance: string;
  checkpointRef: string;
  snapshotWorkspaceRef: string;
  workspaceFileRefs: string[];
  exportBundle: ExportBundleState | null;
};

export type CheckpointComparisonFileState = {
  relativePath: string;
  status: "added" | "removed" | "changed";
  baseRef?: string;
  targetRef?: string;
};

export type CheckpointComparisonState = {
  baseCheckpointId: string;
  baseAlias: string;
  targetCheckpointId: string;
  targetAlias: string;
  comparedFileCount: number;
  changedCount: number;
  addedCount: number;
  removedCount: number;
  summary: string;
  files: CheckpointComparisonFileState[];
};

export type CollectionSummaryState = {
  id: string;
  kind: "raw" | "canonical";
  sourceRef: string;
  timeBucket: string;
  timeRangeLabel: string;
  entryCount: number;
  contentHash: string;
  collectionRef: string;
};

export type CollectionEntryState = {
  id: string;
  sourceRef: string;
  eventTime: string;
  ingestedAt: string;
  contentHash: string;
  preview?: string;
  entryPathRef: string;
  blobRef?: string;
  blobPathRef?: string;
};

export type CollectionDetailState = {
  id: string;
  kind: "raw" | "canonical";
  sourceRef: string;
  timeBucket: string;
  timeRangeLabel: string;
  entryCount: number;
  contentHash: string;
  collectionRef: string;
  entryShardRef: string;
  notes?: string;
  entries: CollectionEntryState[];
};

export type BlobDetailState = {
  id: string;
  blobPathRef: string;
  byteLength: number;
  lineCount: number;
  contentText: string;
};

export type ExchangeAdapterState = {
  id: string;
  name: string;
  kind: string;
  mode: string;
  definitionRef: string;
  supportsLive: boolean;
  supportsPaper: boolean;
  supportsBacktest: boolean;
  capabilities: string[];
  notes?: string;
};

export type EvaluationRunSummaryState = {
  id: string;
  kind: string;
  status: string;
  headline: string;
  summary: string;
  createdAt: string;
  adapterRef: string;
  adapterName: string;
  collectionRefs: string[];
  netPnl: number;
  tradeCount: number;
  positionCount: number;
  pathRef: string;
};

export type EvaluationTradeState = {
  symbol: string;
  side: string;
  entryTime: string;
  exitTime: string;
  entryPrice: number;
  exitPrice: number;
  netPnl: number;
};

export type EvaluationRunDetailState = {
  id: string;
  kind: string;
  status: string;
  headline: string;
  summary: string;
  createdAt: string;
  adapterRef: string;
  adapterName: string;
  collectionRefs: string[];
  grossPnl: number;
  feeCost: number;
  slippageCost: number;
  modelCost: number;
  netPnl: number;
  tradeCount: number;
  positionCount: number;
  pathRef: string;
  equityCurve: EquityPoint[];
  trades: EvaluationTradeState[];
  notes: string[];
};

export type WorkspaceDocumentBacklinkState = {
  label: string;
  pathRef: string;
  category: WorkspaceCatalogCategory | "reference";
  reason: string;
};

export type WorkspaceDocumentState = {
  pathRef: string;
  format: "json" | "ndjson" | "text";
  byteLength: number;
  lineCount: number;
  contentText: string;
  backlinks: WorkspaceDocumentBacklinkState[];
};

export type WorkspaceCatalogEntry = {
  id: string;
  category: WorkspaceCatalogCategory;
  label: string;
  description: string;
  pathRef: string;
};

export type WorkspaceSearchResultState = {
  id: string;
  category: WorkspaceCatalogEntry["category"];
  label: string;
  description: string;
  pathRef: string;
  matchKind: "metadata" | "content";
  excerpt?: string;
};

export type ImportSummaryState = {
  id: string;
  importedAt: string;
  sourceBundleRef: string;
  importRef: string;
  workspaceRef: string;
  checkpointRef: string;
  policyId: string;
  sanitized: boolean;
};

export type ImportPreflightCheckState = {
  id: string;
  severity: ImportPreflightSeverity;
  label: string;
  detail: string;
};

export type ImportPreflightState = {
  status: ImportPreflightStatus;
  summary: string;
  checks: ImportPreflightCheckState[];
};

export type ImportDetailState = ImportSummaryState & {
  bundleRef: string;
  workspaceFileRefs: string[];
  preflight: ImportPreflightState;
};

export type ImportComparisonState = {
  importId: string;
  sourceBundleRef: string;
  comparedFileCount: number;
  changedCount: number;
  addedCount: number;
  removedCount: number;
  summary: string;
  files: CheckpointComparisonFileState[];
};

export type IngestSourceEntryInput = {
  kind: "raw" | "canonical";
  sourceRef: string;
  eventTime: string;
  ingestedAt: string;
  preview?: string;
  bodyText?: string;
};

export type IngestSourceEntryResult = {
  collectionId: string;
  collectionRef: string;
  entryId: string;
  entryShardRef: string;
  timeBucket: string;
  entryCount: number;
  blobId?: string;
  createdCollection: boolean;
};

export type ImportBundleState = {
  importId: string;
  importedAt: string;
  sourceBundleRef: string;
  importRef: string;
  workspaceRef: string;
  checkpointRef: string;
  policyId: string;
  sanitized: boolean;
};

export type OperationSummaryState = {
  id: string;
  kind: string;
  scope: "live" | "workspace";
  status: OperationStatus;
  summary: string;
  details: string;
  createdAt: string;
  operationRef: string;
  relatedRefs: string[];
};

export type OperationRelatedDocumentState = {
  pathRef: string;
  label: string;
  description: string;
  category: string;
  resolved: boolean;
};

export type OperationDetailState = {
  id: string;
  kind: string;
  scope: "live" | "workspace";
  status: OperationStatus;
  summary: string;
  details: string;
  createdAt: string;
  operationRef: string;
  relatedRefs: string[];
  relatedDocuments: OperationRelatedDocumentState[];
  unresolvedRefs: string[];
};

export type LiveSessionState = {
  id: string;
  label: string;
  startedAt: string;
  status: string;
  pathRef: string;
};

export type LiveEvaluationSummaryState = {
  id: string;
  headline: string;
  createdAt: string;
  pathRef: string;
  evidenceRefs: string[];
};

export type WorkspaceSummary = {
  artifactId: string;
  slug: string;
  liveLaneLabel: string;
  currentCheckpointAlias: string;
  exportPolicyLabel: string;
};

export type AssetInspectorState = {
  workspaceRoot: string;
  strategyRef: string;
  liveLaneRef: string;
  currentCheckpointRef: string;
  exportPolicyRef: string;
  latestExportBundleRef?: string;
  checkpointCount: number;
  exportCount: number;
};

export type RuntimeTopologyWorkspaceRefState = {
  label: string;
  pathRef: string;
};

export type OrchestratorRuntimeState = {
  id: string;
  name: string;
  mode: OrchestratorMode;
  pathRef: string;
  notes: string[];
  topologyRefs: {
    agentsRef: string;
    environmentsRef: string;
    sessionsRef: string;
    liveLaneRef: string;
  };
};

export type AgentRuntimeState = {
  id: string;
  name: string;
  kind: string;
  definitionRef: string;
  providerMode: string;
  preferredProviders: string[];
  environmentRef: string;
  environmentName: string;
  workspaceRefs: RuntimeTopologyWorkspaceRefState[];
};

export type EnvironmentRuntimeState = {
  id: string;
  name: string;
  kind: string;
  definitionRef: string;
  capabilities: string[];
  notes?: string;
};

export type RuntimeTopologyState = {
  orchestrator: OrchestratorRuntimeState;
  agents: AgentRuntimeState[];
  environments: EnvironmentRuntimeState[];
};

export type StrategyActiveIndexState = {
  orchestratorRef: string;
  liveLaneRef: string;
  currentCheckpointRef: string;
  exportPolicyRef: string;
};

export type StrategyIndexesState = {
  checkpointsRef: string;
  agentsRef: string;
  environmentsRef: string;
  adaptersRef: string;
  collectionsRef: string;
  evaluationsRef: string;
  importsRef: string;
  operationsRef: string;
  sessionsRef: string;
};

export type WorkspaceIndexState = {
  schemaVersion: string;
  active: StrategyActiveIndexState;
  indexes: StrategyIndexesState;
  agentCount: number;
  environmentCount: number;
  adapterCount: number;
  collectionCount: number;
  evaluationCount: number;
  operationCount: number;
  sessionCount: number;
};

export type LiveContextState = {
  runtimeStatusRef: string;
  dashboardRef: string;
  decisionsRef: string;
  memoryRef: string;
  positionsRef: string;
  ordersRef: string;
  memoryNotes: string[];
  sessions: LiveSessionState[];
  evaluationSummaries: LiveEvaluationSummaryState[];
  positionEventCount: number;
  orderEventCount: number;
};

export type ExportInspectorState = {
  policyId: string;
  description: string;
  latestBundle: ExportBundleState | null;
};

File: src-tauri/src/models/workspace.rs
use serde::{Deserialize, Serialize};

use super::{
    DecisionEntry, EquityPoint, ExposurePoint, LaneEventState, LiveOrder, LivePosition,
    MetricCardData, PricePoint, ProviderStatus, TradingMode,
};

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointSummary {
    pub id: String,
    pub alias: String,
    pub r#type: String,
    pub type_tone: String,
    pub summary: String,
    pub created_at: String,
    pub performance: String,
    pub path_ref: String,
    pub export_bundle_ref: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportBundleState {
    pub export_id: String,
    pub created_at: String,
    pub policy_id: String,
    pub checkpoint_ref: String,
    pub workspace_ref: String,
    pub bundle_ref: String,
    pub included_refs: Vec<String>,
    pub excluded_paths: Vec<String>,
    pub sanitized: bool,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointDetailState {
    pub id: String,
    pub alias: String,
    pub r#type: String,
    pub type_tone: String,
    pub summary: String,
    pub created_at: String,
    pub performance: String,
    pub checkpoint_ref: String,
    pub snapshot_workspace_ref: String,
    pub workspace_file_refs: Vec<String>,
    pub export_bundle: Option<ExportBundleState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointComparisonFileState {
    pub relative_path: String,
    pub status: String,
    pub base_ref: Option<String>,
    pub target_ref: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointComparisonState {
    pub base_checkpoint_id: String,
    pub base_alias: String,
    pub target_checkpoint_id: String,
    pub target_alias: String,
    pub compared_file_count: usize,
    pub changed_count: usize,
    pub added_count: usize,
    pub removed_count: usize,
    pub summary: String,
    pub files: Vec<CheckpointComparisonFileState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CollectionSummaryState {
    pub id: String,
    pub kind: String,
    pub source_ref: String,
    pub time_bucket: String,
    pub time_range_label: String,
    pub entry_count: usize,
    pub content_hash: String,
    pub collection_ref: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CollectionEntryState {
    pub id: String,
    pub source_ref: String,
    pub event_time: String,
    pub ingested_at: String,
    pub content_hash: String,
    pub preview: Option<String>,
    pub entry_path_ref: String,
    pub blob_ref: Option<String>,
    pub blob_path_ref: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CollectionDetailState {
    pub id: String,
    pub kind: String,
    pub source_ref: String,
    pub time_bucket: String,
    pub time_range_label: String,
    pub entry_count: usize,
    pub content_hash: String,
    pub collection_ref: String,
    pub entry_shard_ref: String,
    pub notes: Option<String>,
    pub entries: Vec<CollectionEntryState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportSummaryState {
    pub id: String,
    pub imported_at: String,
    pub source_bundle_ref: String,
    pub import_ref: String,
    pub workspace_ref: String,
    pub checkpoint_ref: String,
    pub policy_id: String,
    pub sanitized: bool,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreflightCheckState {
    pub id: String,
    pub severity: String,
    pub label: String,
    pub detail: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreflightState {
    pub status: String,
    pub summary: String,
    pub checks: Vec<ImportPreflightCheckState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportDetailState {
    pub id: String,
    pub imported_at: String,
    pub source_bundle_ref: String,
    pub import_ref: String,
    pub workspace_ref: String,
    pub checkpoint_ref: String,
    pub policy_id: String,
    pub sanitized: bool,
    pub bundle_ref: String,
    pub workspace_file_refs: Vec<String>,
    pub preflight: ImportPreflightState,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportComparisonState {
    pub import_id: String,
    pub source_bundle_ref: String,
    pub compared_file_count: usize,
    pub changed_count: usize,
    pub added_count: usize,
    pub removed_count: usize,
    pub summary: String,
    pub files: Vec<CheckpointComparisonFileState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlobDetailState {
    pub id: String,
    pub blob_path_ref: String,
    pub byte_length: usize,
    pub line_count: usize,
    pub content_text: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationRelatedDocumentState {
    pub path_ref: String,
    pub label: String,
    pub description: String,
    pub category: String,
    pub resolved: bool,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationDetailState {
    pub id: String,
    pub kind: String,
    pub scope: String,
    pub status: String,
    pub summary: String,
    pub details: String,
    pub created_at: String,
    pub operation_ref: String,
    pub related_refs: Vec<String>,
    pub related_documents: Vec<OperationRelatedDocumentState>,
    pub unresolved_refs: Vec<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceDocumentBacklinkState {
    pub label: String,
    pub path_ref: String,
    pub category: String,
    pub reason: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceDocumentState {
    pub path_ref: String,
    pub format: String,
    pub byte_length: usize,
    pub line_count: usize,
    pub content_text: String,
    pub backlinks: Vec<WorkspaceDocumentBacklinkState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSearchResultState {
    pub id: String,
    pub category: String,
    pub label: String,
    pub description: String,
    pub path_ref: String,
    pub match_kind: String,
    pub excerpt: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IngestSourceEntryInput {
    pub kind: String,
    pub source_ref: String,
    pub event_time: String,
    pub ingested_at: String,
    pub preview: Option<String>,
    pub body_text: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IngestSourceEntryResult {
    pub collection_id: String,
    pub collection_ref: String,
    pub entry_id: String,
    pub entry_shard_ref: String,
    pub time_bucket: String,
    pub entry_count: usize,
    pub blob_id: Option<String>,
    pub created_collection: bool,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportBundleState {
    pub import_id: String,
    pub imported_at: String,
    pub source_bundle_ref: String,
    pub import_ref: String,
    pub workspace_ref: String,
    pub checkpoint_ref: String,
    pub policy_id: String,
    pub sanitized: bool,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSummary {
    pub artifact_id: String,
    pub slug: String,
    pub live_lane_label: String,
    pub current_checkpoint_alias: String,
    pub export_policy_label: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AssetInspectorState {
    pub workspace_root: String,
    pub strategy_ref: String,
    pub live_lane_ref: String,
    pub current_checkpoint_ref: String,
    pub export_policy_ref: String,
    pub latest_export_bundle_ref: Option<String>,
    pub checkpoint_count: usize,
    pub export_count: usize,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyActiveIndexState {
    pub orchestrator_ref: String,
    pub live_lane_ref: String,
    pub current_checkpoint_ref: String,
    pub export_policy_ref: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyIndexesState {
    pub checkpoints_ref: String,
    pub agents_ref: String,
    pub environments_ref: String,
    pub collections_ref: String,
    pub imports_ref: String,
    pub operations_ref: String,
    pub sessions_ref: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceIndexState {
    pub schema_version: String,
    pub active: StrategyActiveIndexState,
    pub indexes: StrategyIndexesState,
    pub agent_count: usize,
    pub environment_count: usize,
    pub collection_count: usize,
    pub operation_count: usize,
    pub session_count: usize,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationSummaryState {
    pub id: String,
    pub kind: String,
    pub scope: String,
    pub status: String,
    pub summary: String,
    pub details: String,
    pub created_at: String,
    pub operation_ref: String,
    pub related_refs: Vec<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceCatalogEntryState {
    pub id: String,
    pub category: String,
    pub label: String,
    pub description: String,
    pub path_ref: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveSessionState {
    pub id: String,
    pub label: String,
    pub started_at: String,
    pub status: String,
    pub path_ref: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveEvaluationSummaryState {
    pub id: String,
    pub headline: String,
    pub created_at: String,
    pub path_ref: String,
    pub evidence_refs: Vec<String>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveContextState {
    pub dashboard_ref: String,
    pub decisions_ref: String,
    pub memory_ref: String,
    pub positions_ref: String,
    pub orders_ref: String,
    pub memory_notes: Vec<String>,
    pub sessions: Vec<LiveSessionState>,
    pub evaluation_summaries: Vec<LiveEvaluationSummaryState>,
    pub position_event_count: usize,
    pub order_event_count: usize,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportInspectorState {
    pub policy_id: String,
    pub description: String,
    pub latest_bundle: Option<ExportBundleState>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BootstrapState {
    pub mode: TradingMode,
    pub automation_status: String,
    pub status_note: Option<String>,
    pub workspace: WorkspaceSummary,
    pub asset_inspector: AssetInspectorState,
    pub workspace_index: WorkspaceIndexState,
    pub live_context: LiveContextState,
    pub export_inspector: ExportInspectorState,
    pub providers: Vec<ProviderStatus>,
    pub metrics: Vec<MetricCardData>,
    pub price_series: Vec<PricePoint>,
    pub equity_series: Vec<EquityPoint>,
    pub exposure_series: Vec<ExposurePoint>,
    pub positions: Vec<LivePosition>,
    pub orders: Vec<LiveOrder>,
    pub lane_events: Vec<LaneEventState>,
    pub decisions: Vec<DecisionEntry>,
    pub checkpoints: Vec<CheckpointSummary>,
    pub collections: Vec<CollectionSummaryState>,
    pub imports: Vec<ImportSummaryState>,
    pub operations: Vec<OperationSummaryState>,
    pub document_catalog: Vec<WorkspaceCatalogEntryState>,
}
